File: backend/internal/service/translate.go
// Package service implements application services with Django parity.
package service

import (
	"container/list"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"
)

// SupportedLanguages mirrors apps.core.translate.SUPPORTED_LANGUAGES.
var SupportedLanguages = []string{"en", "ru", "de", "fr", "es", "pt", "zh-CN", "ja", "ko"}

// ExaSearchLanguages mirrors apps.core.translate.EXA_SEARCH_LANGUAGES.
var ExaSearchLanguages = []string{"en", "ru", "de", "fr", "es", "zh-CN", "ja", "ko"}

// SourceLanguages mirrors apps.core.translate.SOURCE_LANGUAGES.
var SourceLanguages = map[string]string{
	"europe_pmc": "en", "openalex": "en", "crossref": "en", "pubmed": "en",
	"arxiv": "en", "doaj": "en", "pmc": "en", "core": "en", "biorxiv": "en",
	"medrxiv": "en", "dblp": "en", "hal": "fr", "zenodo": "en", "iacr": "en",
	"exa":   "multi",
	"cinii": "ja", "coaj": "zh-CN", "sciengine": "zh-CN", "sciopen": "zh-CN",
	"cyberleninka": "ru", "mathnet": "ru", "scielo": "es", "redalyc": "es",
	"korea_science": "ko", "persee": "fr", "open_edition": "fr",
	"revistas_csic": "es", "medknow": "en", "dergipark": "en", "hrcak": "en",
	"ajol": "en",
}

const translateCacheSize = 512

// Translate provides cross-lingual query translation with GoogleTranslator
// (no key) and MyMemory fallback, LRU-cached. Parity with
// apps.core.translate.
type Translate struct {
	client *http.Client
	cache  *lruCache
}

// NewTranslate builds the translation service.
func NewTranslate() *Translate {
	return &Translate{
		client: &http.Client{Timeout: 15 * time.Second},
		cache:  newLRUCache(translateCacheSize),
	}
}

// DetectLanguage classifies a short query by script (parity with
// apps.core.translate._detect_language).
func DetectLanguage(text string) string {
	for _, r := range text {
		if r >= 0x0400 && r <= 0x04FF {
			return "ru"
		}
	}
	for _, r := range text {
		switch {
		case r >= 0x4E00 && r <= 0x9FFF:
			return "zh-CN"
		case (r >= 0x3040 && r <= 0x309F) || (r >= 0x31F0 && r <= 0x31FF):
			return "ja"
		case r >= 0xAC00 && r <= 0xD7AF:
			return "ko"
		}
	}
	return "en"
}

// TranslateQuery translates a short query into targetLang, falling back to
// MyMemory when Google fails and to the original when both fail. Results are
// cached (parity with apps.core.translate.translate_query).
func (t *Translate) TranslateQuery(ctx context.Context, query, targetLang string) string {
	if strings.TrimSpace(query) == "" {
		return query
	}
	source := DetectLanguage(query)
	if source == targetLang {
		return query
	}
	cacheKey := query + "\x00" + targetLang
	if cached, ok := t.cache.get(cacheKey); ok {
		return cached
	}
	queryText := strings.TrimSpace(query)
	if len(queryText) > 500 {
		queryText = queryText[:500]
	}

	if result, err := t.google(ctx, queryText, targetLang); err == nil && strings.TrimSpace(result) != "" {
		result = strings.TrimSpace(result)
		t.cache.put(cacheKey, result)
		return result
	}
	if result, err := t.myMemory(ctx, queryText, targetLang); err == nil && strings.TrimSpace(result) != "" {
		result = strings.TrimSpace(result)
		t.cache.put(cacheKey, result)
		return result
	}
	t.cache.put(cacheKey, query)
	return query
}

func (t *Translate) google(ctx context.Context, query, targetLang string) (string, error) {
	endpoint := "https://translate.googleapis.com/translate_a/single"
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, endpoint, nil)
	if err != nil {
		return "", err
	}
	params := url.Values{}
	params.Set("client", "gtx")
	params.Set("sl", "auto")
	params.Set("tl", targetLang)
	params.Set("dt", "t")
	params.Set("q", query)
	req.URL.RawQuery = params.Encode()

	resp, err := t.client.Do(req)
	if err != nil {
		return "", err
	}
	defer func() { _ = resp.Body.Close() }()
	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("google translate status %d", resp.StatusCode)
	}
	body, err := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
	if err != nil {
		return "", err
	}
	var segments [][]any
	if err := json.Unmarshal(body, &segments); err != nil {
		return "", err
	}
	var parts []string
	for _, seg := range segments {
		if len(seg) > 0 {
			if s, ok := seg[0].(string); ok {
				parts = append(parts, s)
			}
		}
	}
	return strings.Join(parts, ""), nil
}

func (t *Translate) myMemory(ctx context.Context, query, targetLang string) (string, error) {
	endpoint := "https://api.mymemory.translated.net/get"
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, endpoint, nil)
	if err != nil {
		return "", err
	}
	params := url.Values{}
	params.Set("q", query)
	params.Set("langpair", "auto|"+targetLang)
	req.URL.RawQuery = params.Encode()

	resp, err := t.client.Do(req)
	if err != nil {
		return "", err
	}
	defer func() { _ = resp.Body.Close() }()
	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("mymemory status %d", resp.StatusCode)
	}
	var payload struct {
		ResponseData struct {
			TranslatedText string `json:"translatedText"`
		} `json:"responseData"`
	}
	if err := json.NewDecoder(io.LimitReader(resp.Body, 1<<20)).Decode(&payload); err != nil {
		return "", err
	}
	return payload.ResponseData.TranslatedText, nil
}

// GetSourceQueryLanguage returns the primary query language for a source key.
func GetSourceQueryLanguage(sourceKey string) string {
	if lang, ok := SourceLanguages[sourceKey]; ok {
		return lang
	}
	return "en"
}

// TranslateQueryForSource translates the query into the source's language.
func (t *Translate) TranslateQueryForSource(ctx context.Context, query, sourceKey string) string {
	target := GetSourceQueryLanguage(sourceKey)
	if target == "multi" {
		return query
	}
	return t.TranslateQuery(ctx, query, target)
}

// ExpandQueryForExa translates the query into every Exa target language.
func (t *Translate) ExpandQueryForExa(ctx context.Context, query string) map[string]string {
	results := make(map[string]string, len(ExaSearchLanguages))
	for _, lang := range ExaSearchLanguages {
		if translated := t.TranslateQuery(ctx, query, lang); translated != "" {
			results[lang] = translated
		}
	}
	return results
}

// ExpandSearchTerms returns deduplicated cross-lingual translations of query,
// excluding the original. Parity with apps.core.translate.expand_search_terms.
func (t *Translate) ExpandSearchTerms(ctx context.Context, query string) []string {
	seen := map[string]bool{strings.TrimSpace(query): true}
	var terms []string
	for _, lang := range []string{"en", "ru", "de", "fr", "es"} {
		translated := t.TranslateQuery(ctx, query, lang)
		normalized := strings.TrimSpace(translated)
		key := strings.ToLower(normalized)
		if normalized != "" && !seen[key] {
			terms = append(terms, normalized)
			seen[key] = true
		}
	}
	return terms
}

type lruEntry struct {
	key   string
	value string
}

// lruCache is a small thread-safe LRU string cache.
type lruCache struct {
	mu      sync.Mutex
	max     int
	entries map[string]*list.Element
	order   *list.List
}

func newLRUCache(max int) *lruCache {
	return &lruCache{
		max:     max,
		entries: make(map[string]*list.Element, max),
		order:   list.New(),
	}
}

func (c *lruCache) get(key string) (string, bool) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if el, ok := c.entries[key]; ok {
		c.order.MoveToFront(el)
		if entry, ok2 := el.Value.(*lruEntry); ok2 {
			return entry.value, true
		}
	}
	return "", false
}

func (c *lruCache) put(key, value string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if el, ok := c.entries[key]; ok {
		if entry, ok2 := el.Value.(*lruEntry); ok2 {
			entry.value = value
		}
		c.order.MoveToFront(el)
		return
	}
	el := c.order.PushFront(&lruEntry{key: key, value: value})
	c.entries[key] = el
	if c.order.Len() > c.max {
		oldest := c.order.Back()
		if oldest != nil {
			c.order.Remove(oldest)
			if entry, ok2 := oldest.Value.(*lruEntry); ok2 {
				delete(c.entries, entry.key)
			}
		}
	}
}

File: apps/core/translate.py
"""
Cross-lingual query translation for scholarly search.

Uses deep-translator (GoogleTranslator, no API key) with MyMemory fallback.
Translates short search queries (2-10 words) to target languages so each
source connector receives queries in its primary language, and Exa can search
across multiple languages.
"""

from __future__ import annotations

from functools import lru_cache

import structlog
from deep_translator import GoogleTranslator, MyMemoryTranslator

logger = structlog.get_logger(__name__)

SUPPORTED_LANGUAGES = ("en", "ru", "de", "fr", "es", "pt", "zh-CN", "ja", "ko")

SOURCE_LANGUAGES: dict[str, str] = {
    # API connectors
    "europe_pmc": "en",
    "openalex": "en",
    "crossref": "en",
    "pubmed": "en",
    "arxiv": "en",
    "doaj": "en",
    "pmc": "en",
    "core": "en",
    "biorxiv": "en",
    "medrxiv": "en",
    "dblp": "en",
    "hal": "fr",
    "zenodo": "en",
    "iacr": "en",
    "exa": "multi",
    # HTML connectors
    "cinii": "ja",
    "coaj": "zh-CN",
    "sciengine": "zh-CN",
    "sciopen": "zh-CN",
    "cyberleninka": "ru",
    "mathnet": "ru",
    "scielo": "es",
    "redalyc": "es",
    "korea_science": "ko",
    "persee": "fr",
    "open_edition": "fr",
    "revistas_csic": "es",
    "medknow": "en",
    "dergipark": "en",
    "hrcak": "en",
    "ajol": "en",
}

EXA_SEARCH_LANGUAGES = ("en", "ru", "de", "fr", "es", "zh-CN", "ja", "ko")

_CACHE_MAXSIZE = 512


def _detect_language(text: str) -> str:
    """
    Return the heuristic language code for a short query string.

    Args:
        text: A short query string to classify.

    Returns:
        An ISO 639-1 language code (``"en"``, ``"ru"``, ``"zh-CN"``,
        ``"ja"``, or ``"ko"``).

    """
    for c in text:
        if "Ѐ" <= c <= "ӿ":
            return "ru"
    for c in text:
        if "一" <= c <= "鿿":
            return "zh-CN"
        if "぀" <= c <= "ゟ" or "ㇰ" <= c <= "ヿ":
            return "ja"
        if "가" <= c <= "힯":
            return "ko"
    return "en"


@lru_cache(maxsize=_CACHE_MAXSIZE)
def translate_query(query: str, target_lang: str) -> str:
    """
    Translate a short scholarly query to *target_lang*.

    Uses GoogleTranslator (free, no key) with MyMemory fallback.
    Results are LRU-cached to avoid redundant API calls across search jobs.

    Args:
        query: The source query string (typically 2-10 words).
        target_lang: ISO 639-1 target language code.

    Returns:
        The translated query, or the original if both translators fail.

    """
    if not query or not query.strip():
        return query

    source_lang = _detect_language(query)
    if source_lang == target_lang:
        return query

    query_text = query.strip()[:500]

    try:
        result = GoogleTranslator(source="auto", target=target_lang).translate(
            text=query_text,
        )
        if result and result.strip():
            return result.strip()
    except (ValueError, RuntimeError, ConnectionError) as exc:
        logger.warning(
            "GoogleTranslator failed %s->%s: %s",
            source_lang,
            target_lang,
            exc,
        )

    try:
        result = MyMemoryTranslator(source="auto", target=target_lang).translate(
            text=query_text,
        )
        if result and result.strip():
            return result.strip()
    except (ValueError, RuntimeError, ConnectionError) as exc:
        logger.warning(
            "MyMemoryTranslator failed %s->%s: %s",
            source_lang,
            target_lang,
            exc,
        )

    logger.error("All translators failed for query=%s target=%s", query, target_lang)
    return query


def get_source_query_language(source_key: str) -> str:
    """
    Return the primary query language for a given source key.

    Args:
        source_key: The connector source key (e.g. ``"cinii"``, ``"hal"``).

    Returns:
        An ISO 639-1 language code, or ``"multi"`` for multi-language sources.

    """
    return SOURCE_LANGUAGES.get(source_key, "en")


def translate_query_for_source(query: str, source_key: str) -> str:
    """
    Translate *query* to the primary language of *source_key*.

    Args:
        query: The original user query.
        source_key: The connector source key.

    Returns:
        The translated query, or the original if the source is multi-language.

    """
    target = get_source_query_language(source_key)
    if target == "multi":
        return query
    return translate_query(query, target)


def expand_query_for_exa(query: str) -> dict[str, str]:
    """
    Translate the query into all Exa target languages.

    Args:
        query: The original user query.

    Returns:
        A dict mapping language codes to translated query strings.
        Used by ExaConnector to make one API call per language.

    """
    results: dict[str, str] = {}
    for lang in EXA_SEARCH_LANGUAGES:
        translated = translate_query(query, lang)
        if translated:
            results[lang] = translated
    return results


def expand_search_terms(query: str) -> list[str]:
    """
    Return deduplicated cross-lingual translations of *query*.

    Used by SearchService to expand DB search beyond the original language.

    Args:
        query: The original user query.

    Returns:
        A list of unique translated query strings (excluding the original).

    """
    seen: set[str] = {query.strip().casefold()}
    terms: list[str] = []
    for lang in ("en", "ru", "de", "fr", "es"):
        translated = translate_query(query, lang)
        normalized = translated.strip().casefold()
        if normalized and normalized not in seen:
            terms.append(translated.strip())
            seen.add(normalized)
    return terms
